app/web/features/auth/notifications/PushNotificationPermission.tsx

```diff
@@ -0,0 +1,161 @@
+import { arrayBufferToBase64 } from "utils/arrayBufferToBase64";
+import makeStyles from "utils/makeStyles";
+
+const useStyles = makeStyles((theme) => ({
+  alert: {
+    marginBottom: theme.spacing(3),
+    marginTop: theme.spacing(2),
+  },
+  status: {
+    marginBottom: theme.spacing(2),
+  },
+  titleBox: {
+    display: "flex",
+    alignItems: "center",
+  },
+}));
+
+export default function PushNotificationPermission({
+  className,
+}: {
+  className: string;
+}) {
+  const { t } = useTranslation(AUTH);
+  const classes = useStyles();
+
+  const [permission, setPermission] = useState(Notification.permission);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const hasPermission = permission === "granted";
+
+  const onPermissionGranted = async () => {
+    // If permission is granted, subscribe the user to push notifications
+    try {
+      // Check if service workers and push notifications are supported
+      if ("serviceWorker" in navigator && "PushManager" in window) {
+        const registration = await navigator.serviceWorker.register(
+          "/service-worker.js",
+          { scope: "/" }
+        );
+
+        const existingPushSubscription =
+          await registration.pushManager.getSubscription();
+        const p256dhKey = existingPushSubscription?.getKey("p256dh");
+        const { vapidPublicKey } = await getVapidPublicKey();
+
+        if (existingPushSubscription && p256dhKey) {
+          const publicKey = arrayBufferToBase64(p256dhKey);
+
+          /**
+           * The purpose of this check is to ensure that the push subscription is correctly authenticated with the server’s VAPID key.
+           * If the client’s p256dh key no longer matches the server’s vapidPublicKey, then the subscription is unsubscribed and needs
+           * to be re-registered to ensure the security and validity of the Web Push connection.
+           */
+          if (publicKey !== vapidPublicKey) {
+            await existingPushSubscription.unsubscribe();
+          } else {
+            return;
+          }
+        }
+
+        // Subscribe to push notifications using the PushManager API
+        const subscription: PushSubscription =
+          await registration.pushManager.subscribe({
+            userVisibleOnly: true,
+            applicationServerKey: vapidPublicKey,
+          });
+
+        await registerPushNotificationSubscription(subscription);
+      } else {
+        setErrorMessage(
+          t("notification_settings.push_notifications.error_unsupported")
+        );
+      }
+    } catch (error) {
+      setErrorMessage(
+        t("notification_settings.push_notifications.error_generic")
+      );
+    }
+  };
+
+  const turnPushNotificationsOn = async () => {
+    if (permission === "default") {
+      const result = await Notification.requestPermission();
+      setPermission(result);
+
+      if (result === "granted") {
+        await onPermissionGranted();
+      }
```

Comment:
What if I already granted permission (so the browser can show notifications), but I turned it off and want to re-turn on?

Then I think you should have `permission` equal to `granted` above and this function takes the error path instead?

Reply:
I'm not totally following this.

I think maybe you're saying there should be a has permission but turned off push notifications state, but right now there's just the `Notification.permission` state?

I mean, I guess we could do that but we'd need to either store the user settings in `localStorage`, on the user object in the DB or something like that.

Do you mean you don't want turning the switch off to equal permission denied (if they granted it first then turned it off after)?

Reply:
Hmm so when a user unsubscribes from push notifications, they need to manually reset the permission through the browser settings to bring it back to default. There's no programmatic way to automatically set the permission back to default after unsubscribing unfortunately.

So only way around with would be to persist some kinda "push turned off" value if they ever enabled them.

I guess this would be either storing some user settings in that user's row in the database, or `localStorage` or something like that.

Do you know offhand how y'all persist user info like that? Is that with `usePersistedState`?

Reply:
Ohhhh, I see what you are saying. I mean in order to know if the user is currently subscribed to push notifs on this browser, I suppose we just have to check if they have a subscription on the service worker, right?

Reply:
Okay this is done now!